utils/bloom_filter.cc

```diff
@@ -74,6 +75,39 @@ bool bloom_filter::is_present(const bytes_view& key) {
     return is_present(make_hashed_key(key));
 }
 
+void bloom_filter::fold(const size_t new_num_bits) {
+    const auto curr_num_bits = _bitset.size();
+    if (curr_num_bits % new_num_bits != 0) {
+        seastar::on_internal_error(filterlog, std::format("cannot fold down bloom filter as the new number of bits({}) is not a factor of existing bitset size({})", new_num_bits, curr_num_bits));
+        return;
+    }
+
+    // The folding will happen inplace (i.e) the first new_num_bits of the current
+    // bitmap will be updated with the folded filter's bits. A bit b[i] where 'i'
+    // belonging to [0, new_num_bits), will be set if any of its input bits :
+    // b[i], b[i + new_num_bits], b[i + 2*new_num_bits]... is set in the current bitmap.
+    for (uint64_t new_bitset_idx = 0; new_bitset_idx < new_num_bits; new_bitset_idx++) {
+        if (_bitset.test(new_bitset_idx)) {
+            // the target bit is already set - no need to look at other input bits
+            continue;
+        }
+
+        // set the target bit if any of its input bit is set
+        auto old_bitset_idx = new_bitset_idx + new_num_bits;
+        while (old_bitset_idx < curr_num_bits) {
+            if (_bitset.test(old_bitset_idx)) {
+                _bitset.set(new_bitset_idx);
+                break;
```

Comment:
if we'll keep the 64-bit alignment, we can do bitwise OR on 64-bits, no? 1 bit at a time sounds expensive.

Reply:
Yes, current code has terrible performance. Much better to do bitwise or.